helm-app-operator/pkg/helm/helm.go

```diff
@@ -442,8 +493,25 @@ func (c installer) tillerRendererForCR(r *unstructured.Unstructured) *tiller.Rel
 	return tiller.NewReleaseServer(env, internalClientSet, false)
 }
 
-func releaseName(r *unstructured.Unstructured) string {
-	return fmt.Sprintf("%s-%s", operatorName, r.GetName())
+func getReleaseName(r *unstructured.Unstructured) string {
+	status := v1alpha1.StatusFor(r)
+	if status.Release != nil {
```

Comment:
@joelanford The implementation seems fine to me but I'm thinking we should try our best to avoid relying on Status for our control logic. That seems to be the upstream convention as well.
 
https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#spec-and-status
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/principles.md#api
> Object status must be 100% reconstructable by observation. Any history kept must be just an optimization and not required for correct operation.

The `valueResourceUID` is useful in letting us verify the correct owner of a pre-existing release.
But we still end up storing the release name in the Status to look up again in the future.

Right now with this implementation what happens if:
-  We deploy a new release with no release name, which generates a new unique release name e.g "foo"
    ```Go
      c.installRelease(tiller, r.GetNamespace(), "", chart, config)
    ```
- Fail to update the status with `updatedRelease=foo`
- Retry the reconcile. See that `status.Release` is empty from our last failed attempt.
- Retry deploying the release with no release name
    ```Go
      c.installRelease(tiller, r.GetNamespace(), "", chart, config)
    ```
- Does that give us a new deployed release with a new unique name `updatedRelease=bar`?

Ideally we could make the UUID part of the release name e.g `cr.name+UUID` to make it fully unique which avoids the problem of saving the release name since we can reconstruct it just by looking at the CR. And that solves the ownership problem.
But there's the issue of the 53 char limit. I know UUID's are 36 chars so we could stay within the limit but I'm wondering how human readable do release names need to be?

Reply:
Yeah, you're right. That would result in a separate new release.

Thanks for linking the conventions and principles. That's good info!

I did think about using UUIDs in the release name, but there are some downsides:
* As you point out, UUIDs would take up 36 of the 53 characters, leaving only 17 if we wanted to include the CR name. That may be cutting it a little too close for comfort.
* Release names end up in the generated resource names for all helm chart resources, so users would end up with really long names for everything managed by the helm operator.

We could do something like `cr.Name + last_N_chars(cr.UUID)`, where N is < 36. Assuming the UUIDs are uniformally random hex characters, the probability of collision would likely be sufficiently low at N=8.

If we want better human readability, another option could be that we add the release name to the CR (somewhere in `metadata` or `spec` maybe?) prior to any attempt to invoke the release installation, update, etc. That way a CR never results in any deployed release until it has been updated with a release name.